### DISCUSSION TITLE
TMS Optimization for normal or tangential E-field components

### DISCUSSION THREAD

**Discussion by fs210** [simnibs/simnibs#437] [8. Simulation workflows]
Hi SIMNIBS experts,

I am interested in using the TMS Optimization python code to create TMS coil poses which are optimized for the normal and tangential components of the induced E-field at a target site (e.g. X,Y,Z coordinates in the DLPFC). I understand that the E-field magnitude may be optimized at the target site by simply not including the 'target_direction' attribute, however, how can I specify either the normal or tangential E-field components?

In the attribute descriptions, I see the following:

> target_direction: None/[] or (3x1) list/array (Python/MATLAB) of floats, optional
> Desctiption: Direction of the electric field to be optimized. If None (Python) or [] (MATLAB), will optimize the elctric field magnitude. Default: optimize field magnitude
> Example: Python/MATLAB
> opt.target_direction = [1, 0, 0]

Does the assignment of 'opt.target_direction' need to be specified for the target site's normal vector components? If so, how should I go about finding the normal or tangential components? If they are not vector components, can someone please define what the 3x1 list represents?

Thanks for any help!

**Comment by TorgeW** [comment]
'opt.target_direction' is the direction you want your E field to be at the target. 
This can be the normal vector at the target, in this case you would select the position that maximizes the current in or out flow into the cortex at the target position.

**Reply by fs210** [reply]
Thanks @TorgeW! Is there an easy way to figure out what those vector components would be for either the normal or tangential components (I'm guessing the normal vector would be easier). Is there some python code which relays what the nearest E-field normal vector (pointing outward) would be at a specific target location? I understand that simply multiplying this vector by -1 would point the direction inward.

**Reply by fs210** [reply]
Hi again,

Just following up. Is there some python code which relays what the nearest E-field normal vector would be at a specific target location/coordinate?

**Reply by TorgeW** [reply]
Is what you want the normal to the gray matter surface, if yes you can get it by doing this:

```
from simnibs import mesh_io

m = mesh_io.read_msh("path/to/mesh")
gm_surf = m.crop_mesh(tags=[1002])
normals = msh.triangle_normals()
```

You could then just select a normal at the position you want.

**Reply by fs210** [reply]
Ok. How can I implement this in the TMS Optimize python code? It would be useful if the Optimize code simply had a choice for which E-field form to optimize (magnitude, normal, tangential, etc.) at the target site.

As a second question, if I understand correctly, the code will produce a matsimnibs matrix. How can this be implemented/loaded in SimNIBS? Is there a simple way to relay what the source position and source orientation should be in image space coordinates (e.g. X, Y, Z --> 40, 29, 22; i, j, k --> -0.25, -0.5, -0.5 (pointing to the left, posteriorly, and tangential to the head)?

**Reply by TorgeW** [reply]
You first have to determine the desired electric field direction (direction vector) using the normals of the triangles (generated as described above) that are close to your target position. Based on the normal close to your target you can also generate a tangential direction vector. You will then set this direction vector in the setup of the optimization (`opt.target_direction = direction_vector`). 
The optimization will optimize the magnitude if target_direction is set to None, if target_direction is not None and you have set a direction vector, it will optimize for the E field components corresponding to the target_direction. The reason that there is no automatic setting for the different options, is that the normal inside a 3D volume is ambiguous. 
https://simnibs.github.io/simnibs/build/html/documentation/opt_struct/tmsoptimize.html 
https://simnibs.github.io/simnibs/build/html/tutorial/tms_optimization.html

The optimization will indeed result in an affine transformation matrix (4x4).
When setting up a simulation you will add a position. Instead of specifying the centre and pos_ydir you can use the affine matrix received from the optimization. 
`pos.matsimnibs = matsimnibs_from_optimization`
https://simnibs.github.io/simnibs/build/html/documentation/sim_struct/position.html#position-doc
